### PR TITLE
SDA package with vector field

### PR DESCRIPTION
**Pull request by liyu711**
Deleted and added the original SDA package
Added the vector field method for moving obstacles only
Add new value to obstacle: type "moving" & "stationary"
New method get_force in VectorMath
New method about repulsive_force and attractive_force in obstacle_map

### CODE REVIEW

File: SDAPackage/SDA/constants.py [hunk 1]
```diff
@@ -4,7 +4,7 @@ class Constants(object):
     """
 
     STATIONARY_OBSTACLE_SAFETY_RADIUS = 30
-    MOVING_OBSTACLE_SAFETY_RADIUS = 30
+    MOVING_OBSTACLE_SAFETY_RADIUS = 90
```
Comment (DarkmatterVale) [inline SDAPackage/SDA/constants.py]: Please do not modify the SDA package contents.

File: SDAPackage/SDA/stationary_obstacle.py [hunk 2]
```diff
@@ -1,51 +0,0 @@
-from SDA import Obstacle
```
Comment (DarkmatterVale) [inline SDAPackage/SDA/stationary_obstacle.py]: Do not modify SDAPackage contents.

File: SDAPackage/SDA/obstacle_map.py [hunk 3]
```diff
@@ -92,6 +92,16 @@ def make_encompassing_circle(self, first_obstacle, second_obstacle, final_obstac
             new_z = second_obstacle.get_point()[2]
         return StationaryObstacle(np.array([new_x, new_y, new_z]), new_radius, new_height)
 
+    def is_moving_obstacle_in_path(self, obstacle):
```
Comment (DarkmatterVale) [inline SDAPackage/SDA/obstacle_map.py]: Do not modify SDA package contents.

File: SDAPackage/SDA/mass.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from SDA import *
```
Comment (DarkmatterVale) [inline SDAPackage/SDA/mass.py]: Do not modify SDA package contents.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import numpy
+from math import atan2, cos, sin, pi, sqrt
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: PEP8 requires only 1 import per line.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from SDAWithVectorField import *
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: Please avoid ```*``` imports.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+class ObstacleMap(object):
+	def __init__(self, drone_point, fly_zones):
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: Please add a blank line above the init method declaration.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		self.obstacles = numpy.array([])
+		self.drone = Drone(drone_point, numpy.array([]))
+		self.fly_zones = FlightBoundariesContainer(fly_zones)
+		self.repulsive_forces = numpy.array([0,0,0])
+		self.attractive_force = numpy.array([0,0,0])
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py:13]: You do not need to declare the variables here; instead, you can just call the ```reset_attractive_force()``` and ```reset_repulsive_forces()``` methods here.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		self.detection_radius = 0
+
+	def add_obstacle(self, obstacle):
+		if self.obstacles.size != 0:
+			final_obstacles = numpy.hstack([self.obstacles, obstacle])
+			self.obstacles = final_obstacles
+		else:
+			self.obstacles = numpy.array([obstacle])
+
+	def add_force(self):
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: This method looks good, but we should change the name to be more descriptive to its function. How does ```update_repulsive_forces()``` sound?

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		for obstacle in self.obstacles:
+			if obstacle.get_type() == "moving":
+				force = VectorMath.get_force(self.drone.get_point(), obstacle.get_point())
+				self.repulsive_forces = numpy.sum([self.repulsive_forces, force], axis=0)
+
+	def set_attractive_force(self):
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: Again, we need a name change I think. What about ```update_attractive_force()```?

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+	def reset_attractive_force(self):
+		self.attractive_force = numpy.array([0,0,0])
+
+	def reset_repulsive_forces(self):
+		self.repulsive_forces = numpy.array([0,0,0])
+
+	def get_avoidance_vector(self):
+		repulsive_forces = self.repulsive_forces
+		attractive_force = self.attractive_force
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: Do you need to make these variables here or can you complete the calculation with the class variable?

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+	def get_repulsive_forces(self):
+		return self.repulsive_forces
+
+	def add_waypoint(self, waypoint):
+		"""
+        Add a waypoint to the drone
+
+        :param waypoint: The waypoint to add
+        :type waypoint: Numpy Array
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: Fix spacing issue here.

File: SDAPackageWithVectorField/SDAWithVectorField/vector_math.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    @staticmethod
+    def get_force(vector_one, vector_two):
+        """
+        return the force vector between two points
+        """
+        distance_vector = np.subtract(vector_one, vector_two)
+        force_vector = np.array([1.0/pow(distance_vector[0],2),1.0/pow(distance_vector[1],2),1.0/pow(distance_vector[2],2)])
+        return force_vector
+
+
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/vector_math.py]: Remove these extra lines.

File: SDAPackageWithVectorField/SDAWithVectorField/vector_math.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        Returns the vector rejection between two vectors
+        """
+        projection_vector = VectorMath.get_vector_projection(vector_one, vector_two)
+
+        return np.subtract(vector_one, projection_vector)
+    
+    @staticmethod
+    def get_force(vector_one, vector_two):
+        """
+        return the force vector between two points
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/vector_math.py]: First word of the method comment should be capitalized.

File: SDAPackageWithVectorField/SDAWithVectorField/vector_math.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        distance_vector = np.subtract(vector_one, vector_two)
+        force_vector = np.array([1.0/pow(distance_vector[0],2),1.0/pow(distance_vector[1],2),1.0/pow(distance_vector[2],2)])
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/vector_math.py]: Please fix the spacing issues here.

File: SDAPackageWithVectorField/setup.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
+# This is the setup file for pip
+from setuptools import setup, find_packages
+import os, sys
+from os import path
+
+setup(
+    name = 'SDAWtihVectorField',
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/setup.py]: Please fix the typo with the name.

File: SDAPackageWithVectorField/setup.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+    version = '0.0.2',
+
+    description = 'Object detection and Avoidance for SUAS Competition',
+
+    url = 'https://github.com/FlintHill/SUAS-Competition',
+
+    author = 'Vale Tolpegin',
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/setup.py]: I am not the author of this package. You should use your name and email address.

File: tests/my_tests/test_obstacle_map_vectorfield.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+		self.obstacle_map1.set_attractive_force()
+		result = self.obstacle_map1.get_avoidance_vector()
+
+		repulsive_forces = VectorMath.get_force(self.test_drone_point, self.obstacle1.get_point())
+		attractive_force = VectorMath.get_force(self.test_drone_point, self.waypoint)
+		avoidance_vector = numpy.sum([attractive_force, repulsive_forces], axis=0)
+
+		self.assertEqual(result[0], avoidance_vector[0])
+		
+
```
Comment (DarkmatterVale) [inline tests/my_tests/test_obstacle_map_vectorfield.py]: Please remove extra lines here.

File: tests/unit_tests/test_obstacle_map.py [hunk 19]
```diff
@@ -358,3 +358,17 @@ def test_get_path_distance(self):
...
+		self.obstacle_map1.reset_obstacles()
+		test_obstacle = MovingObstacle(numpy.array([50,0,2]),10)
+		drone_point = numpy.array([0,0,2])
+		test_waypoint = numpy.array([70,0,2])
+		self.obstacle_map1.add_waypoint(test_waypoint)
+		self.obstacle_map1.add_obstacle(test_obstacle)
+		self.assertTrue(self.obstacle_map1.is_moving_obstacle_in_path(test_obstacle))
+
+	
+
```
Comment (DarkmatterVale) [inline tests/unit_tests/test_obstacle_map.py]: Please remove extra lines here.

File: tests/unit_tests/test_sda_vector_math.py [hunk 20]
```diff
@@ -1,7 +1,7 @@
 import math
 import numpy
 import unittest
-from SDA import VectorMath
+from SDAWithVectorField import VectorMath
```
Comment (DarkmatterVale) [inline tests/unit_tests/test_sda_vector_math.py]: You should create a separate vector math test case for your package, and leave the other unit tests in operation.

File: tests/unit_tests/test_sda_vector_math.py [hunk 21]
```diff
@@ -62,3 +62,21 @@ def test_get_vector_rejection(self):
...
+        force_one = VectorMath.get_force(test_vector1, test_vector2)
+        force_two = VectorMath.get_force(test_vector2, test_vector3)
+
+        print(force_one)
+        print(force_two)
+        self.assertEqual(force_one[0],force_two[0])
+        self.assertEqual(force_one[1],force_two[1])
+        self.assertEqual(force_one[2],force_two[2])
+
+
```
Comment (DarkmatterVale) [inline tests/unit_tests/test_sda_vector_math.py]: Remove extra lines.

**Review by DarkmatterVale — CHANGES_REQUESTED** [review]

File: SDAPackageWithVectorField/SDAWithVectorField/moving_obstacle.py [hunk 22]
```diff
@@ -1,5 +1,5 @@
-from SDA import Obstacle
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/moving_obstacle.py:1]: Did you move this file from SDAPackage to the new SDA system?

File: SDAPackageWithVectorField/SDAWithVectorField/moving_obstacle.py [hunk 23]
```diff
@@ -21,13 +21,16 @@ def __init__(self, point, radius):
         :param radius: The radius of the obstacle
         :type radius: Float
         """
-        super(MovingObstacle, self).__init__(point, Constants.MOVING_OBSTACLE_SAFETY_RADIUS)
+        super(MovingObstacle, self).__init__(point)
 
         self.radius = radius
+        self.type = "moving"
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/moving_obstacle.py]: As we talked about in class, please you class type as the distinguisher between stationary and moving, not a string.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py (hunk 6 above)
Comment (DarkmatterVale) [reply SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: As we talked about in class, this is alright for the main file.

File: SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py (hunk 5 above)
Comment (DarkmatterVale) [reply SDAPackageWithVectorField/SDAWithVectorField/obstacle_map.py]: This still needs to be addressed.

File: SDAPackageWithVectorField/SDAWithVectorField/stationary_obstacle.py [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+        return self.radius + self.get_safety_radius()
+
+    def get_height(self):
+        """
+        The height of the obstacle
+        """
+        return self.height
+
+    def get_type(self):
+        return self.type
```
Comment (DarkmatterVale) [inline SDAPackageWithVectorField/SDAWithVectorField/stationary_obstacle.py]: Same comment about the type issue.

File: install.sh [hunk 25]
```diff
@@ -22,3 +22,6 @@ sudo -H python -m pip install ./EigenFit -U
 sudo -H python -m pip install ./ImgProcessingCLI -U
 sudo -H python -m pip install ./SUASSystem -U
 sudo -H python -m pip install ./SDAPackage -U
+sudo -H python -m pip install ./SDAPackageWithVectorField -U
+
+
```
Comment (DarkmatterVale) [inline install.sh]: Please remove the extra lines.

File: tests/unit_tests/test_obstacle_map.py [hunk 26]
```diff
@@ -358,3 +358,13 @@ def test_get_path_distance(self):
 		self.paths1 = self.obstacle_map1.generate_possible_paths(self.test_obstacle1)
 		self.min_path1 = self.obstacle_map1.get_min_path(self.paths1)
 		self.assertEqual(self.obstacle_map1.get_path_distance(self.min_path1), 236.44394938110665)
+
+	def test_is_moving_obstacle_in_path(self):
```
Comment (DarkmatterVale) [inline tests/unit_tests/test_obstacle_map.py]: We do not use the current SDA system for moving obstacles anymore. Please update your comment accordingly.

**Review by DarkmatterVale — CHANGES_REQUESTED** [review]
Getting much closer!

We need to fix the Travis tests as well as the comments.

**Comment by DarkmatterVale** [comment]
Also, just so everyone is on the same page, this PR will stay open until we are ready to merge the first version of this code in (when the simulator is completed and tested).

File: SDAPackageWithVectorField/SDAWithVectorField/Work need to be done [hunk 27]
```diff
@@ -0,0 +1,6 @@
+2018.1.7 assigned
+due around 2018.1.10
+	#Simulator for obstacle map
+	#Avoid from the edge
+	#Type() instead of get_type()
+
```
Comment (jvillemare) [inline SDAPackageWithVectorField/SDAWithVectorField/Work need to be done:6]: This seems like a notes file. This shouldn't be committed, and instead kept somewhere in the Google Drive.

File: SDAPackageWithVectorField/SDAWithVectorField/flight_boundaries_container.py [hunk 28]
```diff
@@ -0,0 +1,26 @@
+from SDAWithVectorField import *
+import numpy
+
+class FlightBoundariesContainer(object):
```
Comment (jvillemare) [inline SDAPackageWithVectorField/SDAWithVectorField/flight_boundaries_container.py:4]: All classes should have block comments below the signature describing their purpose.

File: SDAPackageWithVectorField/SDAWithVectorField/flight_boundary.py [hunk 29]
```diff
@@ -0,0 +1,48 @@
+from matplotlib.path import Path
+
+class FlightBoundary(object):
+    """
+    Provides an interface for flight zone boundaries
```
Comment (jvillemare) [inline SDAPackageWithVectorField/SDAWithVectorField/flight_boundary.py:5]: Class descriptions should are sentences, so they should have periods.

File: SDAPackageWithVectorField/SDAWithVectorField/flight_boundaries_container.py (hunk 28 above)
Comment (jvillemare) [reply SDAPackageWithVectorField/SDAWithVectorField/flight_boundaries_container.py:4]: This applies to all other files in this PR, as multiple classes are missing block comments.

**Review by jvillemare — CHANGES_REQUESTED** [review]

**Pull request closed without merging**